android/app/src/main/java/com/zulipmobile/notifications/FCMPushNotifications.kt

```diff
@@ -65,129 +71,156 @@ internal fun onReceived(context: Context, conversations: ConversationMap, mapDat
         } else {
-            builder.setContentTitle(fcmMessage.sender.fullName)
-        }
-        builder.setContentText(fcmMessage.content)
-        if (fcmMessage.recipient is Recipient.Stream) {
-            val (stream, topic) = fcmMessage.recipient
-            val displayTopic = "$stream > $topic"
-            builder.setSubText("Message on $displayTopic")
+            builder.setSmallIcon(R.drawable.zulip_notification)
         }
-        fetchBitmap(sizedURL(context, fcmMessage.sender.avatarURL, 64f))
-            ?.let { builder.setLargeIcon(it) }
-        builder.setStyle(Notification.BigTextStyle().bigText(fcmMessage.content))
-    } else {
-        val numConversations = context.resources.getQuantityString(
-            R.plurals.numConversations, conversations.size, conversations.size)
-        builder.setContentTitle("$totalMessagesCount messages in $numConversations")
-        builder.setContentText("Messages from ${TextUtils.join(",", nameList)}")
-        val inboxStyle = Notification.InboxStyle(builder)
-        inboxStyle.setSummaryText(numConversations)
-        buildNotificationContent(conversations, inboxStyle)
-        builder.setStyle(inboxStyle)
+        setStyle(NotificationCompat.InboxStyle()
+            .setSummaryText(realmUri)
+        )
+        setGroup(groupKey)
+        setGroupSummary(true)
+        setAutoCancel(true)
     }
 
-    try {
-        ShortcutBadger.applyCount(context, totalMessagesCount)
-    } catch (e: Exception) {
-        Log.e(TAG, "BADGE ERROR: $e")
-    }
-
-    builder.setWhen(fcmMessage.timeMs)
-    builder.setShowWhen(true)
-
-    val vPattern = longArrayOf(0, 100, 200, 100)
-    // NB the DEFAULT_VIBRATE flag below causes this to have no effect.
-    // TODO: choose a vibration pattern we like, and unset DEFAULT_VIBRATE.
-    builder.setVibrate(vPattern)
+    return builder
+}
 
-    builder.setDefaults(Notification.DEFAULT_VIBRATE or Notification.DEFAULT_LIGHTS)
+private fun updateNotification(
+    context: Context, fcmMessage: MessageFcmMessage) {
+    val user = Person.Builder().setName("You").build()
+    val sender = Person.Builder()
+        .setName(fcmMessage.sender.fullName)
+        .setIcon(IconCompat.createWithBitmap(fetchBitmap(fcmMessage.sender.avatarURL)))
+        .build()
+
+    var isGroupConversation = true
+    val title = when (fcmMessage.recipient) {
+        is Recipient.Stream -> "#${fcmMessage.recipient.stream} > ${fcmMessage.recipient.topic}"
+        // TODO use proper title for GroupPM, we will need
+        // to have a way to get names of PM users here.
+        is Recipient.GroupPm -> context.getString(R.string.group_pm, fcmMessage.recipient.pmUsers)
+        is Recipient.Pm -> {
+            isGroupConversation = false
+            fcmMessage.sender.fullName
+        }
+    }
+    val groupKey = "${fcmMessage.identity?.realmUri.toString()};${fcmMessage.identity?.userId}"
+    val conversationKey = "$groupKey;$title"
```

Comment:
And let's pull this out as its own helper function too, this one taking the `FcmMessage`. This helper's job is:
* The result should be constant for a given conversation.
* It should be distinct for distinct conversations.

That points at a few improvements we can make, relative to the solution of using the user-facing title as it's defined above:
* For type `Recipient.Pm`, use `sender.id` instead of `sender.fullName`. That way (a) it doesn't get messed up if they change their name, (b) there's no possibility of the key getting confused because the delimiter character appeared in the data.
  * Before relying on `sender.id`, we should make it non-null, the same way as `identity` as discussed above; our notes say this also dates to server v1.8.0.
  * If we didn't reliably have `sender.id`, then `sender.email` would be a next-best solution.
* For `Recipient.GroupPm`, `recipient.pmUsers` is perfect. Best to skip the `getString`, in fact, so this internal data structure's workings aren't complicated by interactions with a translated user-facing string.
* For `Recipient.Stream`... ideally we'd use the stream ID instead of the stream name, but it looks like we don't currently have that in the data. For now, I guess we can settle for a TODO comment in the code saying to switch to stream ID. And for a delimiter, let's use the character `\x00`, for the same reason as we do that in `keyFromNarrow` in `narrow.js`:
  ```js
      // '\x00' is the one character not allowed in Zulip stream names.
      // (See `check_stream_name` in zulip.git:zerver/lib/streams.py.)
      topic: (streamName, topic) => `topic:s:${streamName}\x00${topic}`,
  ```
* Oh also, looking at `keyFromNarrow` reminds me: for these conversation keys let's also have a different fixed prefix for each of the three types of recipient. That way, each of the three different ways we format a conversation key always has a distinct prefix, so it's easy to be sure that they'll never accidentally collide with each other.

Reply:
done, using `\u0000` instead of `\x00`, kotlin was not letting me use `\x`, I hope that's fine

Reply:
Sure -- that should be another way (and I guess the only way, in Kotlin) of spelling the same thing.